src/mscorlib/src/System/Runtime/InteropServices/NativeLibrary.cs

```diff
@@ -0,0 +1,365 @@
+        /// <param name="paths">A flags value which contains the locations that should be searched during module load.</param>
+        /// <param name="result">A <see cref="NativeLibrary"/> instance that represents the loaded module,
+        /// or <see langword="null"/> if the module could not be found. Once loaded, a module cannot be unloaded.</param>
+        /// <returns><see langword="true"/> if the module was found and loaded, otherwise <see langword="false"/>.</returns>
+        /// <remarks>
+        /// There are some behavioral differences between this method and <see cref="DllImportAttribute"/>. For example, the
+        /// value provided in <paramref name="name"/> cannot be the display name of a managed assembly. Additionally, if
+        /// <paramref name="caller"/> is not specified, then <paramref name="paths"/> cannot have the flags values
+        /// <see cref="DllImportSearchPath.LegacyBehavior"/> or <see cref="DllImportSearchPath.AssemblyDirectory"/>.
+        /// If <paramref name="caller"/> is specified and <paramref name="paths"/> is <see cref="DllImportSearchPath.LegacyBehavior"/>
+        /// and an assembly-level <see cref="DefaultDllImportSearchPathsAttribute"/> is present, the assembly-level default
+        /// <see cref="DllImportSearchPath"/> value will be used instead.
+        /// This method may throw an exception in certain cases, such as if the module was located but is invalid
+        /// for the current architecture.
+        /// </remarks>
+        public static bool TryLoad(string name, Assembly caller, DllImportSearchPath paths, out NativeLibrary result)
+        {
+            // Parameter validation
+
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            RuntimeAssembly callerAsRuntimeAssembly = caller as RuntimeAssembly;
+            if (caller != null && callerAsRuntimeAssembly == null)
+            {
+                throw new ArgumentException(
+                    paramName: nameof(caller),
+                    message: SR.Argument_MustBeRuntimeAssembly);
+            }
+
+            // CLR's module loader uses the low byte of the flags enum for assembly-relative flags,
+            // so this can't be specified unless a calling assembly has also been specified. In this
+            // case, we require application-level or global flags to be specified, which are in the
+            // high three bytes of the flags enum.
+
+            if (caller == null && (((int)paths & ~0xFF) == 0 || ((int)paths & 0xFF) != 0))
+            {
+                // If no assembly provided, caller cannot specify LegacyBehavior or AssemblyDirectory.
+                // If no assembly provided, caller must specify 
+                // TODO: What should the exception message say?
+                throw new Exception("Must specify DllImportSearchPaths if assembly not specified.");
+            }
+
```

Comment:
In addition to the low byte being reserved, I think it's important to check that the enum value being passed in is one we understand.  Unmapped values will lead to differing behaviors on Windows and non-Windows systems, and we should endeavor to prevent that from happening in new API.

Reply:
I don't know how we could do that without interfering with the scenario. The only values understood by non-Windows platforms are `AssemblyDirectory` and `LegacyBehavior`. We can't eagerly fail on other values being passed in because it would block the scenario below.

```cs
NativeLibrary lib;
if (NativeLibrary.TryLoad("foo.dll", null, System32, out lib))
{
  // We're on Windows, use OS-provided foo.dll.
}
else if (NativeLibrary.TryLoad("foo.so", typeof(me).Assembly, AssemblyDirectory, out lib))
{
  // We're on non-Windows, use our local equivalent.
}
```

We don't want the call to `TryLoad("foo.dll", ...)` to throw since it's really nothing more than a glorified OS check.

Reply:
We spoke offline about this. We can throw as long as the caller has the ability to conditionally guard the call. For example, the guarded call site could look like this.

```cs
NativeLibrary lib;
if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
{
    if (!NativeLibrary.TryLoad("foo.dll", ..., System32, out lib))
    {
        throw new Exception("Expected this DLL to be found.");
    }
}
else
{
    if (!NativeLibrary.TryLoad("bar.so", ..., Assembly, out lib))
    {
        throw new Exception("Expected this DLL to be found.");
    }
}
```

This pattern would also be more resilient to DLL poisoning attacks because the OS version check is performed up front, allowing the caller to pass the search paths appropriate for the OS.